src/jdk.compiler/share/classes/com/sun/tools/javac/comp/ThisEscapeAnalyzer.java

```diff
@@ -0,0 +1,1475 @@
+ * An indirect reference means the reference refers to the 'this' instance we are tracking through
+ * at least one level of indirection.
+ *
+ * <p>
+ * Currently we do not attempt to explicitly track references stored in fields (for future study).
+ *
+ * <p>
+ * A few notes on this implementation:
+ * <ul>
+ *  <li>We "execute" constructors and track where the 'this' reference goes as the constructor executes.
+ *  <li>We use a very simplified flow analysis that you might call a "flood analysis", where the union
+ *      of every possible code branch is taken.
+ *  <li>A "leak" is defined as the possible passing of a subclassed 'this' reference to code defined
+ *      outside of the current compilation unit.
+ *  <ul>
+ *      <li>In other words, we don't try to protect the current compilation unit from itself.
+ *      <li>For example, we ignore private constructors because they can never be directly invoked
+ *          by external subclasses, etc. However, they can be indirectly invoked by other constructors.
+ *  </ul>
+ *  <li>If a constructor invokes a method defined in the same compilation unit, and that method cannot
+ *      be overridden, then our analysis can safely "recurse" into the method.
+ *  <ul>
+ *      <li>When this occurs the warning displays each step in the stack trace to help in comprehension.
+ *  </ul>
+ *  <li>We assume that native methods do not leak.
+ *  <li>We don't try to follow {@code super()} invocations; that's for the superclass analysis to handle.
+ *  </ul>
+ */
+class ThisEscapeAnalyzer extends TreeScanner {
+
+    private final Names names;
+    private final Types types;
+    private final Log log;
+    private       Lint lint;
+
+// These fields are scoped to the entire COMPILATION UNIT
+
+    /** Maps symbols of all methods to their corresponding declarations.
+     */
+    private final Map<Symbol, MethodInfo> methodMap = new LinkedHashMap<>();
+
+    /** Contains symbols of fields and methods that have warnings suppressed.
+     */
+    private final Set<Symbol> suppressed = new HashSet<>();
+
+    /** The declaring class of the constructor we're currently analyzing.
+     *  This is the 'this' type we're trying to detect leaks of.
+     */
+    private JCClassDecl targetClass;
+
+    /** Snapshots of {@link #callStack} where possible 'this' escapes occur.
+     */
+    private ArrayList<DiagnosticPosition[]> warningList = new ArrayList<>();
+
+// These fields are scoped to the CONSTRUCTOR BEING ANALYZED
+
+    /** The declaring class of the "invoked" method we're currently analyzing.
+     *  This is either the analyzed constructor or some method it invokes.
+     */
+    private JCClassDecl methodClass;
+
+    /** The current "call stack" during our analysis. The first entry is some method
+     *  invoked from the target constructor; if empty, we're still in the constructor.
+     */
+    private final ArrayDeque<DiagnosticPosition> callStack = new ArrayDeque<>();
```

Comment:
There is a concept of push/popScope and then there's a separate concept of call stack (which is just a list of diagnostic position up to the point). I wonder if this could be better modeled by using a single class e.g. Scope/Frame which has a diagnostic position, plus other useful things. Perhaps it might even be helpful to have a ref set on each scope, so that you don't have to attach a "depth" to each ref - the depth of the ref would be determined by the "scope" in which it appears.

Reply:
> There is a concept of push/popScope and then there's a separate concept of call stack (which is just a list of diagnostic position up to the point). I wonder if this could be better modeled by using a single class e.g. Scope/Frame which has a diagnostic position, plus other useful things.

I think that would be more confusing, because they're really two different animals.

Scopes only have meaning within a single method. They simply serve to bracket the lifetimes of `VarRef` references, following Java curly brace scope. When you "invoke" a method, you put aside the current stack of scopes and start over with an empty scope stack. They don't bridge between methods.

Call stacks are just a list of method+code position therein, and they exist outside of any single method. They have nothing to do with Java scopes defined by curly braces.

> Perhaps it might even be helpful to have a ref set on each scope, so that you don't have to attach a "depth" to each ref - the depth of the ref would be determined by the "scope" in which it appears.

That's another way to skin the same cat... in fact I had a design like that before but then realized it was a lot simpler to just carry around one `RefSet`.